app/services/plugin.service.ts

```diff
@@ -2,139 +2,101 @@ import { Injectable, Inject } from '@angular/core';
 import { PropertiedEntity, PropertyList } from "../components/properties-panel/properties-panel.component";
 
 import * as Type from "../types/types";
-import * as Core from '../models/core'
-import { Program, InterpreterError, Graph, ProgramInput, ProgramOutput } from "../models/plugin"
-import { Context, SandboxService, Script } from "../services/sandbox.service"
-import { FileService } from "../services/files.service"
+import * as Core from '../models/core';
+import { Object as SinapObject } from "../models/object";
+import { Program, InterpreterError, Graph, ProgramInput, ProgramOutput } from "../models/plugin";
+import { Context, SandboxService, Script } from "../services/sandbox.service";
+import { FileService } from "../services/files.service";
 
-class ConcretePropertyList implements PropertyList {
-    constructor(public properties: [string, Type.Type][], private backerObject: any) {
+// TODO:
+// this file has a bunch of calls to 
+// `instanceof` that could probably be encoded in the 
+// type system
 
-    }
-    get(property: string) {
-        return this.backerObject[property];
-    }
-    set(property: string, value: any) {
-        this.backerObject[property] = value;
-    }
-}
 
-
-class PluginPropertyData implements Core.PluginData {
+export class PluginPropertyData implements Core.PluginData {
     backer: any = {};
-    propertyList: PropertyList;
-    constructor(public type: string, types: [string, Type.Type][]) {
-        this.propertyList = new ConcretePropertyList(types, this.backer);
+    object: SinapObject;
+    constructor(public kind: string, t: Type.ClassType) {
+        this.object = new SinapObject(t, this.backer);
     }
 }
 
-class ConcretePlugin implements Core.Plugin {
-    kind = "dfa.sinap.graph-kind";
-    script: Promise<Script>;
+type Definitions = { all: Map<string, Type.Type>, nodes: Map<string, Type.ClassType>, edges: Map<string, Type.ClassType>, graphs: Map<string, Type.ClassType> };
```

Comment:
Please put this on multiple lines, I had to scroll on GitHub while reviewing this to see the whole thing.